content/languages/c/authoring/memory-management-techniques.md

```diff
@@ -6,19 +6,509 @@ sidebar: "language:c"
+destroy_world(actual);
+```
+
+</details>
+
+As this approach very useful for more complex memory structures, a couple of examples can be found in the section on [Two-dimensional arrays](#two-dimensional-arrays).
+
+
 ## Two-dimensional arrays
 
-- N+1 allocations
-- Flat
-- TOC + Flat
+Some kata require the user solution to return a two-dimensional array, for example a 2D matrix, or an array of C-strings. Such scenarios are a bit more complex, because not only does the higher order array have to be properly managed, but all its individual entries as well. The exact approach selected for allocation of such structures depend on the scenario, because different techniques are suitable for square or rectangular arrays, jagged arrays, arrays of null-terminated strings, etc.
+
+
+Memory for 2D arrays can be managed by the test suite or the user solution. As long as the size of the 2D array is known before calling a solution and does not change through the course of calculations, the test suite can choose to perform all necessary allocations and pass the memory to the solution function ready to use. This is a very good approach when working with chessboards, sudokus, matrices and mazes of predetermined sizes, etc. However, in the case that the size of the answer cannot be easily determined beforehand, the technique with clean-up function provided by the user turns out to be helpful. A user provided clean-up function is used in the examples below, but authors can choose to manage the memory in the test suite if it fits the task of their kata.
+
+
+### Naive approach: N+1 allocations
+
+This is the most common yet worst possible approach of using dynamically allocated memory. It tends to be slow, causes excessive memory fragmentation and is usually inferior to available alternatives.
+
+<details>
+
+```c
+char** game_of_life(int generations, char** initial_generation, int* world_h, int* world_w) {
+
+     //allocating memory for the world map, row by row
+    char** world = malloc(sizeof(char*) * world_h);
+    for(int i=0; i < world_h; ++i)
+        world[i] = malloc(world_w);
+    
+    for(int i=0 i < generations; ++i) {
+        //... actual game, which potentially requires additional (re)allocations...
+    }
+
+    //return the final state of the game world to the caller
+    return world;
+}
+
+void destroy_world(char** world, int world_h) {
+    
+    //... deallocate all memory also row by row
+    for(int i=0; i < world_h; ++i)
+        free(world[i]);
+    free(world);
+}
+```
+
+</details>
+
+### Flat array
+
+Very often overlooked, but a very good approach to represent 2D arrays is to store them in a regular, linear array of `T[ ]`. It can be effectively used when bounds between inner arrays can be efficiently determined, for example each row of a matrix has a well known length, rows of a Pascal's triangle have precisely defined, although different, lengths, and string entries are clearly terminated.
```

Comment:
```suggestion
Very often overlooked, but a very good approach to represent 2D arrays is to store them in a regular, linear array of `T[ ]`. It can be effectively used when bounds between inner arrays can be efficiently determined, for example, each row of a matrix has a well-known length, rows of a Pascal's triangle have precisely defined, although different, lengths, and string entries are clearly terminated.
```